packages/@aws-cdk/tmp-toolkit-helpers/src/api/refactoring/index.ts

```diff
@@ -0,0 +1,415 @@
+    super('Ambiguous resource mappings');
+  }
+
+  public paths(): [string[], string[]][] {
+    return this.pairs.map(([a, b]) => [convert(a), convert(b)]);
+
+    function convert(locations: ResourceLocation[]): string[] {
+      return locations.map((l) => l.toPath());
+    }
+  }
+}
+
+/**
+ * This class mirrors the `ResourceLocation` interface from CloudFormation,
+ * but is richer, since it has a reference to the stack object, rather than
+ * merely the stack name.
+ */
+export class ResourceLocation {
+  constructor(readonly stack: BasicStack, readonly logicalResourceId?: string) {
+  }
+
+  public toPath(): string {
+    const stack = this.stack;
+    const resource = stack.template.Resources?.[this.logicalResourceId ?? ''];
+    const result = resource?.Metadata?.['aws:cdk:path'];
+
+    if (result != null) {
+      return result;
+    }
+
+    // If the path is not available, we can use stack name and logical ID
+    return `${stack.stackName}.${this.logicalResourceId}`;
+  }
+}
+
+/**
+ * A mapping between a source and a destination location.
+ */
+export class ResourceMapping {
+  constructor(public readonly source: ResourceLocation, public readonly destination: ResourceLocation) {
+  }
+
+  public toTypedMapping(): TypedMapping {
+    return {
+      // the type is the same in both source and destination,
+      // so we can use either one
+      type: this.getType(this.source),
+      sourcePath: this.source.toPath(),
+      destinationPath: this.destination.toPath(),
+    };
+  }
+
+  private getType(location: ResourceLocation): string {
+    const resource = location.stack.template.Resources?.[location.logicalResourceId ?? ''];
+    return resource?.Type ?? 'Unknown';
+  }
+}
+
+function buildRecord(entries: [string, ResourceLocation][]): Record<string, ResourceLocation[]> {
+  const result: Record<string, ResourceLocation[]> = {};
+
+  for (const [hash, location] of entries) {
+    if (hash in result) {
+      result[hash].push(location);
+    } else {
+      result[hash] = [location];
+    }
+  }
+
+  return result;
+}
+
+export function computeMappings(before: BasicStack[], after: BasicStack[]): ResourceMapping[] {
+  const pairs = removeCommonResources(zip(buildRecord(before.flatMap(index)), buildRecord(after.flatMap(index))));
+
+  // A mapping is considered ambiguous if these two conditions are met:
+  //  1. Both sides have at least one element (otherwise, it's just addition or deletion)
+  //  2. At least one side has more than one element
+  const ambiguousPairs = Object.values(pairs)
+    .filter(([pre, post]) => pre.length > 0 && post.length > 0)
+    .filter(([pre, post]) => pre.length > 1 || post.length > 1);
+
+  if (ambiguousPairs.length > 0) {
+    throw new AmbiguityError(ambiguousPairs);
+  }
+
+  return Object.values(pairs)
+    .filter(([pre, post]) => pre.length === 1 && post.length === 1 && !equalLocations(pre[0], post[0]))
+    .map(([pre, post]) => new ResourceMapping(pre[0], post[0]));
+}
```

Comment:
It's unfortunate that the return type is:

* Either everything is unambiguous and here is the answer; or
* There is at least one ambiguity, so YOU GET NOTHING

Can we not return both the suggested and ambiguous mappings, and make the decision to throw somewhere else?